plugin/forward/dnstap.go

```diff
@@ -0,0 +1,55 @@
+package forward
+
+import (
+	"context"
+	"time"
+
+	"github.com/coredns/coredns/plugin/dnstap"
+	"github.com/coredns/coredns/plugin/dnstap/msg"
+
+	tap "github.com/dnstap/golang-dnstap"
+	"github.com/miekg/dns"
+)
+
+func logToDnstap(ctx context.Context, host, proto string, req, ret *dns.Msg, start time.Time) error {
+	t := dnstap.TapperFromContext(ctx)
+	if t == nil {
+		return nil
+	}
+
+	b := msg.New().Time(start).HostPort(host)
+
+	if proto == "tcp" {
+		b.SocketProto = tap.SocketProtocol_TCP
+	} else {
+		b.SocketProto = tap.SocketProtocol_UDP
+	}
+
+	if t.Pack() {
+		b.Msg(req)
+	}
+
+	m, err := b.ToOutsideQuery(tap.Message_FORWARDER_QUERY)
```

Comment:
consider adding a config option to dnstap plugin which defines which types of dnstap messages should be logged and which ones should not, e.g.
```
dnstap <socket> {
    messages CQ CR FQ FR AR RQ
}
```
Other plugins could send (or not to send) particular dnstap messages based on this configuration

Reply:
I need to think about this

Reply:
This could be done in a separate PR